libraries/AP_RangeFinder/AP_RangeFinder_VL53L1X.cpp

```diff
@@ -0,0 +1,474 @@
+// based on VL53L1_calc_macro_period_us()
+uint32_t AP_RangeFinder_VL53L1X::calcMacroPeriod(uint8_t vcsel_period)
+{
+    // from VL53L1_calc_pll_period_us()
+    // fast osc frequency in 4.12 format; PLL period in 0.24 format
+    uint32_t pll_period_us = ((uint32_t)0x01 << 30) / fast_osc_frequency;
+
+    // from VL53L1_decode_vcsel_period()
+    uint8_t vcsel_period_pclks = (vcsel_period + 1) << 1;
+
+    // VL53L1_MACRO_PERIOD_VCSEL_PERIODS = 2304
+    uint32_t macro_period_us = (uint32_t)2304 * pll_period_us;
+    macro_period_us >>= 6;
+    macro_period_us *= vcsel_period_pclks;
+    macro_period_us >>= 6;
+
+    return macro_period_us;
+}
+
+// read - return last value measured by sensor
+bool AP_RangeFinder_VL53L1X::get_reading(uint16_t &reading_mm)
+{
+    if (!dataReady()) {
+        return false;
+    }
+
+    reading_mm = read_register16(RESULT__FINAL_CROSSTALK_CORRECTED_RANGE_MM_SD0);
+
+    // "apply correction gain"
+    // gain factor of 2011 is tuning parm default (VL53L1_TUNINGPARM_LITE_RANGING_GAIN_FACTOR_DEFAULT)
+    // Basically, this appears to scale the result by 2011/2048, or about 98%
+    // (with the 1024 added for proper rounding).
+    reading_mm = ((uint32_t)reading_mm * 2011 + 0x0400) / 0x0800;
+    
+    write_register(SYSTEM__INTERRUPT_CLEAR, 0x01); // sys_interrupt_clear_range
+    return true;
+}
+
+uint8_t AP_RangeFinder_VL53L1X::read_register(uint16_t reg)
+{
+    uint8_t v = 0;
+    uint8_t b[2] = { uint8_t(reg >> 8), uint8_t(reg & 0xFF) };
+    dev->transfer(b, 2, &v, 1);
+    return v;
+
+}
+
+uint16_t AP_RangeFinder_VL53L1X::read_register16(uint16_t reg)
+{
+    uint16_t v = 0;
+    uint8_t b[2] = { uint8_t(reg >> 8), uint8_t(reg & 0xFF) };
+    dev->transfer(b, 2, (uint8_t *)&v, 2);
+    return be16toh(v);
+}
+
+uint32_t AP_RangeFinder_VL53L1X::read_register32(uint16_t reg)
+{
+    uint32_t v = 0;
+    uint8_t b[2] = { uint8_t(reg >> 8), uint8_t(reg & 0xFF) };
+    dev->transfer(b, 2, (uint8_t *)&v, 4);
+    return be32toh(v);
+}
+
+void AP_RangeFinder_VL53L1X::write_register(uint16_t reg, uint8_t value)
+{
+    uint8_t b[3] = { uint8_t(reg >> 8), uint8_t(reg & 0xFF), value };
+    dev->transfer(b, 3, nullptr, 0);
+}
+
+void AP_RangeFinder_VL53L1X::write_register16(uint16_t reg, uint16_t value)
+{
+    uint8_t b[4] = { uint8_t(reg >> 8), uint8_t(reg & 0xFF), uint8_t(value >> 8), uint8_t(value & 0xFF) };
+    dev->transfer(b, 4, nullptr, 0);
+}
+
+void AP_RangeFinder_VL53L1X::write_register32(uint16_t reg, uint32_t value)
+{
+    uint8_t b[6] = { uint8_t(reg >> 8),
+                     uint8_t(reg & 0xFF),
+                     uint8_t((value >> 24) & 0xFF),
+                     uint8_t((value >> 16) & 0xFF),
+                     uint8_t((value >>  8) & 0xFF),
+                     uint8_t((value)       & 0xFF) };
+    dev->transfer(b, 6, nullptr, 0);
+}
+
+/*
+   update the state of the sensor
+*/
+void AP_RangeFinder_VL53L1X::update(void)
+{
+    if (counter > 0) {
```

Comment:
Needs a semaphore.

Reply:
Hi @OXINARF,
I changed the code and used `timer` to read & update distance measurement, I removed the update function as I think it isn't needed (similar to `AP_RangeFinder_PulsedLightLRF`). 

I can't find any example of using semaphore in reading phase but only in init function. Could you tell me why we need to add semaphore while reading, or any example?

Thank you so much for reviewing my code!

Reply:
Actually that's worse (thanks for noticing that in PulsedLightRF, we need to fix it). The `update` method is called from the main thread, at a known time, the timer runs in a separate thread and shouldn't be updating frontend values asynchronously.

And that's the same reason why the semaphore is needed, you shouldn't read and write the values from separate threads without having synchronization protection.

If you want an example, look at MaxsonarI2CXL or TeraRangerI2C. All I2C drivers should have it by the way, sometimes stuff like that goes unnoticed.